File: src/main/java/com/yupi/airouter/controller/StripeWebhookController.java
package com.yupi.airouter.controller;

import com.stripe.model.Event;
import com.stripe.model.checkout.Session;
import com.yupi.airouter.service.StripePaymentService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

/**
 * Stripe Webhook 控制器
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@RestController
@RequestMapping("/webhook/stripe")
@Tag(name = "Stripe Webhook")
@Slf4j
public class StripeWebhookController {

    @Resource
    private StripePaymentService stripePaymentService;

    /**
     * Stripe Webhook 回调
     * Stripe会在支付状态变更时调用这个接口
     */
    @PostMapping
    @Operation(summary = "Stripe Webhook回调")
    public String handleStripeWebhook(
            @RequestBody String payload,
            @RequestHeader("Stripe-Signature") String sigHeader) {

        log.info("收到Stripe Webhook回调");

        try {
            // 验证 Webhook 签名
            Event event = stripePaymentService.constructWebhookEvent(payload, sigHeader);

            // 处理不同的事件类型
            switch (event.getType()) {
                case "checkout.session.completed":
                    // 支付成功
                    Session session = (Session) event.getDataObjectDeserializer()
                            .getObject().orElse(null);
                    if (session != null) {
                        log.info("支付成功：SessionID {}", session.getId());
                        stripePaymentService.handlePaymentSuccess(session.getId());
                    }
                    break;

                case "checkout.session.expired":
                    // 支付会话过期
                    log.info("支付会话过期");
                    break;

                case "charge.refunded":
                    // 退款
                    log.info("订单退款");
                    break;

                default:
                    log.info("未处理的事件类型：{}", event.getType());
            }

            return "success";
        } catch (Exception e) {
            log.error("处理Stripe Webhook失败", e);
            return "error";
        }
    }
}

File: go-backend/internal/controller/stripe_webhook_controller.go
package controller

import (
	"log"

	"github.com/gin-gonic/gin"
	"github.com/stripe/stripe-go/v84"
	"github.com/yupi/airouter/go-backend/internal/service"
)

type StripeWebhookController struct {
	stripePaymentService *service.StripePaymentService
}

func NewStripeWebhookController(stripePaymentService *service.StripePaymentService) *StripeWebhookController {
	return &StripeWebhookController{
		stripePaymentService: stripePaymentService,
	}
}

func (s *StripeWebhookController) HandleStripeWebhook(c *gin.Context) {
	payloadBytes, err := c.GetRawData()
	if err != nil {
		log.Printf("read stripe webhook payload failed: err=%v", err)
		c.String(200, "error")
		return
	}
	sigHeader := c.GetHeader("Stripe-Signature")
	event, err := s.stripePaymentService.ConstructWebhookEvent(string(payloadBytes), sigHeader)
	if err != nil {
		log.Printf("verify stripe webhook signature failed: err=%v", err)
		c.String(200, "error")
		return
	}
	switch event.Type {
	case "checkout.session.completed":
		sess, parseErr := service.ParseCheckoutSessionFromEvent(event)
		if parseErr != nil {
			log.Printf("parse stripe checkout session failed: err=%v", parseErr)
			c.String(200, "error")
			return
		}
		if _, handleErr := s.stripePaymentService.HandlePaymentSuccess(sess.ID); handleErr != nil {
			log.Printf("handle stripe payment success failed: sessionId=%s err=%v", sess.ID, handleErr)
			c.String(200, "error")
			return
		}
	case "checkout.session.expired", "charge.refunded":
		// 与 Java 保持一致：记录日志，不做额外业务处理
		log.Printf("stripe event received: type=%s", event.Type)
	default:
		log.Printf("stripe event ignored: type=%s", event.Type)
	}
	c.String(200, "success")
}

func IsStripeEvent(event *stripe.Event, eventType string) bool {
	return event != nil && event.Type == stripe.EventType(eventType)
}
